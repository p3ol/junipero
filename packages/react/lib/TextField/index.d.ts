import React, { ComponentPropsWithRef, MutableRefObject } from "react";

export declare type TextFieldRef = {
  innerRef: MutableRefObject<any>;
  inputRef: MutableRefObject<any>;
  value: String;
  valid: Boolean;
  dirty: Boolean;
  focused: Boolean;
  focus: () => void;
  blur: () => void;
  reset: () => void;
  setDirty: (dirty: Boolean) => void;
  isJunipero: Boolean;
};

declare interface TextFieldProps extends ComponentPropsWithRef<any> {
  autoFocus?: Boolean;
  children: React.ReactNode;
  className: String;
  valid: Boolean;
  value?: String;
  disabled: Boolean;
  required?: Boolean;
  type?: String;
  onBlur?: (event: Event) => void;
  onFocus?: (event: Event) => void;
  onChange?: (props: { value: String; valid: Boolean; dirty: Boolean }) => void;
  onValidate?: (
    val: String,
    options: { required?: Boolean; dirty?: Boolean }
  ) => Boolean;
  ref?: MutableRefObject<TextFieldRef | undefined>;
}

declare function TextField(props: TextFieldProps): JSX.Element;

export default TextField;
