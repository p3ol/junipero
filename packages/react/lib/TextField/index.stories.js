import { action } from '@storybook/addon-actions';

import Label from '../Label';
import Abstract from '../Abstract';
import TextField from './index';
import FieldControl from '../FieldControl';

export default { title: 'react/TextField' };

export const basic = () => (
  <TextField placeholder="First name" onChange={action('change')} />
);

export const basicMultiline = () => (
  <TextField
    rows={10}
    label="First name"
    placeholder="First name"
    onChange={action('change')}
  />
);

export const withLabelAndAbstract = () => (
  <FieldControl>
    <Label className="info" htmlFor="firstname">First name</Label>
    <TextField
      id="firstname"
      placeholder="Type your first name"
      onChange={action('change')}
      onValidate={() => false}
    />
    <Abstract className="info">
      Here is a little help writing your name
    </Abstract>
  </FieldControl>
);

export const autoFocused = () => (
  <TextField autoFocus onChange={action('change')} />
);

export const disabled = () => (
  <TextField disabled value="test" />
);

export const withValidation = () => (
  <TextField
    label="Age"
    placeholder="Age"
    onValidate={val => /^[0-9]+$/g.test(val)}
  />
);
