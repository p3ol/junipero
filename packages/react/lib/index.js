// Display
export { default as Abstract } from './Abstract';
export { default as Alert } from './Alert';
export { default as Alerts } from './Alerts';
export { default as AlertsControl } from './AlertsControl';
export { default as Badge } from './Badge';
export { default as BreadCrumb } from './BreadCrumb';
export { default as BreadCrumbItem } from './BreadCrumbItem';
export { default as Button } from './Button';
export { default as Calendar } from './Calendar';
export { default as Card } from './Card';
export { default as Draggable } from './Draggable';
export { default as Dropdown } from './Dropdown';
export { default as DropdownGroup } from './DropdownGroup';
export { default as DropdownItem } from './DropdownItem';
export { default as DropdownMenu } from './DropdownMenu';
export { default as DropdownToggle } from './DropdownToggle';
export { default as Droppable } from './Droppable';
export { default as Label } from './Label';
export { default as List } from './List';
export { default as ListCell } from './ListCell';
export { default as ListColumn } from './ListColumn';
export { default as ListItem } from './ListItem';
export { default as Modal } from './Modal';
export { default as Slider } from './Slider';
export { default as Spinner } from './Spinner';
export { default as Tab } from './Tab';
export { default as Tabs } from './Tabs';
export { default as Tag } from './Tag';
export { default as Toast } from './Toast';
export { default as Toasts } from './Toasts';
export { default as ToastsControl } from './ToastsControl';
export { default as Toggle } from './Toggle';
export { default as Tooltip } from './Tooltip';
export { default as TouchableZone } from './TouchableZone';
export { default as Transition } from './Transition';

// Forms
export { default as CheckboxField } from './CheckboxField';
export { default as CodeField } from './CodeField';
export { default as ColorField } from './ColorField';
export { default as DateField } from './DateField';
export { default as FieldAddon } from './FieldAddon';
export { default as FieldControl } from './FieldControl';
export { default as FieldGroup } from './FieldGroup';
export { default as RadioField } from './RadioField';
export { default as SelectField } from './SelectField';
export { default as TextField } from './TextField';

// Logos
export {
  PooolLogo,
  PooolCompactLogo,
  AccessLogo,
  SubscribeLogo,
  ConnectLogo,
  EngageLogo,
  FlowLogo,
} from './Logo';

// Core
export {
  classNames,
  addClass,
  removeClass,
  mockState,
  isUndefined,
  isNull,
  isArray,
  isObject,
  isDate,
  exists,
  get,
  set,
  omitBy,
  omit,
  pick,
  cloneDeep,
  fromPairs,
  mergeDeep,
  ensureNode,
  findDeep,
  filterDeep,
  ensureMinMax,
  getFloatPrecision,
  formatThousand,
} from '@junipero/core';

// Hooks
export {
  useEventListener,
  useInterval,
  useTimeout,
  useEffectAfterMount,
  useLayoutEffectAfterMount,
} from '@junipero/hooks';

export {
  useAlerts,
  useToasts,
} from './hooks';

// Icons
export {
  Arrows,
  Remove,
  ArrowLeft,
  ArrowRight,
  ArrowDown,
  ArrowUp,
  Time,
  Check,
} from './icons';
