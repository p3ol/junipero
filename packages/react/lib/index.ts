export {
  COLORS,
  ensureNode,
  classNames,
  addClass,
  removeClass,
  mockState,
  exists,
  isUndefined,
  isNull,
  isArray,
  isObject,
  isDate,
  get,
  set,
  omitBy,
  omit,
  pick,
  cloneDeep,
  fromPairs,
  mergeDeep,
  filterDeep,
  findDeep,
  hsva2hsla,
  hsla2hsva,
  hsva2rgba,
  rgba2hex,
  rgba2hsva,
  denormalizeHSLA,
  denormalizeHSVA,
  denormalizeRGBA,
  parseColor,
  stringifyColor,
  addMonths,
  subMonths,
  startOfYear,
  endOfYear,
  startOfMonth,
  endOfMonth,
  startOfDay,
  endOfDay,
  getDaysInMonth,
  closestIndexTo,
  ensureMinMax,
  getFloatPrecision,
  formatThousand,
} from '@junipero/core';

export {
  useEventListener,
  useInterval,
  useTimeout,
  useEffectAfterMount,
  useLayoutEffectAfterMount,
} from '@junipero/hooks';

export {
  default as Abstract,
  type AbstractProps,
} from './Abstract';
export {
  default as Alert,
  type AlertRef,
  type AlertProps,
  type AlertObject,
} from './Alert';
export {
  default as Alerts,
  type AlertsRef,
  type AlertsProps,
} from './Alerts';
export {
  default as AlertsControl,
  type AlertsControlRef,
  type AlertsControlProps,
} from './AlertsControl';
export {
  default as Badge,
  type BadgeRef,
  type BadgeProps,
} from './Badge';
export {
  default as BreadCrumb,
  type BreadCrumbRef,
  type BreadCrumbProps,
} from './BreadCrumb';
export {
  default as BreadCrumbItem,
  type BreadCrumbItemRef,
  type BreadCrumbItemProps,
} from './BreadCrumbItem';
export {
  default as Button,
  type ButtonRef,
  type ButtonProps,
} from './Button';
export {
  default as Calendar,
  type CalendarRef,
  type CalendarProps,
} from './Calendar';
export {
  default as Card,
  type CardRef,
  type CardProps,
} from './Card';
export {
  default as CheckboxField,
  type CheckboxFieldRef,
  type CheckboxFieldProps,
} from './CheckboxField';
export {
  default as CodeField,
  type CodeFieldRef,
  type CodeFieldProps,
} from './CodeField';
export {
  default as ColorField,
  type ColorFieldRef,
  type ColorFieldProps,
} from './ColorField';
export {
  default as DateField,
  type DateFieldRef,
  type DateFieldProps,
} from './DateField';
export {
  default as Draggable,
  type DraggableProps,
} from './Draggable';
export {
  default as Dropdown,
  type DropdownRef,
  type DropdownProps,
} from './Dropdown';
export {
  default as DropdownGroup,
  type DropdownGroupProps,
} from './DropdownGroup';
export {
  default as DropdownItem,
  type DropdownItemProps,
} from './DropdownItem';
export {
  default as DropdownMenu,
  type DropdownMenuRef,
  type DropdownMenuProps,
} from './DropdownMenu';
export {
  default as DropdownToggle,
  type DropdownToggleRef,
  type DropdownToggleProps,
} from './DropdownToggle';
export {
  default as Droppable,
  type DroppableProps,
} from './Droppable';
export {
  default as FieldAddon,
  type FieldAddonRef,
  type FieldAddonProps,
} from './FieldAddon';
export {
  default as FieldControl,
} from './FieldControl';
export {
  default as FieldGroup,
  type FieldGroupRef,
  type FieldGroupProps,
} from './FieldGroup';
export {
  default as Label,
  type LabelProps,
} from './Label';
export {
  default as List,
  type ListRef,
  type ListProps,
} from './List';
export {
  default as ListCell,
  type ListCellRef,
  type ListCellProps,
} from './ListCell';
export {
  default as ListColumn,
  type ListColumnRef,
  type ListColumnProps,
  type ListColumnObject,
} from './ListColumn';
export {
  default as ListItem,
  type ListItemRef,
  type ListItemProps,
} from './ListItem';
export {
  default as Modal,
  type ModalRef,
  type ModalProps,
} from './Modal';
export {
  default as ModalControl,
  type ModalControlRef,
  type ModalControlProps,
} from './ModalControl';
export {
  PooolLogo,
  PooolIconLogo,
  PooolCutoutLogo,
  PooolCutoutIconLogo,
  PooolCompactLogo,
  PooolCompactCutoutLogo,
  AccessLogo,
  SubscribeLogo,
  ConnectLogo,
  EngageLogo,
  FlowLogo,
  JuniperoLogo,
} from './Logo';
export {
  default as RadioField,
  type RadioFieldRef,
  type RadioFieldProps,
} from './RadioField';
export {
  default as SelectField,
  type SelectFieldRef,
  type SelectFieldProps,
} from './SelectField';
export {
  default as Slider,
  type SliderRef,
  type SliderProps,
} from './Slider';
export {
  default as Spinner,
  type SpinnerProps,
} from './Spinner';
export {
  default as Step,
  type StepRef,
  type StepProps,
  type StepObject,
} from './Step';
export {
  default as Stepper,
  type StepperRef,
  type StepperProps,
} from './Stepper';
export {
  default as Tab,
  type TabRef,
  type TabProps,
  type TabObject,
} from './Tab';
export {
  default as Tabs,
  type TabsRef,
  type TabsProps,
} from './Tabs';
export {
  default as Tag,
  type TagRef,
  type TagProps,
} from './Tag';
export {
  default as TextField,
  type TextFieldRef,
  type TextFieldProps,
} from './TextField';
export {
  default as Toast,
  type ToastRef,
  type ToastProps,
  type ToastObject,
} from './Toast';
export {
  default as Toasts,
  type ToastsRef,
  type ToastsProps,
} from './Toasts';
export {
  default as ToastsControl,
  type ToastsControlRef,
  type ToastsControlProps,
} from './ToastsControl';
export {
  default as Toggle,
  type ToggleRef,
  type ToggleProps,
} from './Toggle';
export {
  default as Tooltip,
  type TooltipRef,
  type TooltipProps,
} from './Tooltip';
export {
  default as TouchableZone,
  type TouchableZoneProps,
} from './TouchableZone';
export {
  default as Transition,
  type TransitionProps,
} from './Transition';

export {
  ListContext,
  AlertsContext,
  ToastsContext,
  DropdownContext,
  FieldControlContext,
  ModalContext,
  useList,
  useAlerts,
  useToasts,
  useDropdown,
  useFieldControl,
  useModal,
} from './hooks';

export {
  Arrows,
  Remove,
  ArrowLeft,
  ArrowRight,
  ArrowDown,
  ArrowUp,
  Time,
  Check,
} from './icons';

export type {
  Grow,
  GrowToSize,
  FixedArray,
} from './utils';
