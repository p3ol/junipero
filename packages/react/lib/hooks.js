import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
