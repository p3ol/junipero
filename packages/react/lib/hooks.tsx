import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
