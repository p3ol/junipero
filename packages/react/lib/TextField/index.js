import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState, exists } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const TextField = forwardRef(({
  min,
  max,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val, { required }) => !!val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = value => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = e => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = e => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = e => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = dirty => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        min={min}
        max={max}
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? (() => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
});

TextField.displayName = 'TextField';
TextField.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.string,
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onWheel: PropTypes.func,
  onValidate: PropTypes.func,
};

export default TextField;
