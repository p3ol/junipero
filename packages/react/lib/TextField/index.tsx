import {
  type ChangeEvent,
  type FocusEvent,
  type RefObject,
  type WheelEvent,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, exists } from '@junipero/core';

import type {
  FieldContent,
  JuniperoRef,
  SpecialComponentPropsWithRef,
} from '../types';
import { mockState } from '../utils';
import { useFieldControl } from '../hooks';

export declare interface TextFieldRef extends JuniperoRef {
  dirty: boolean;
  focused: boolean;
  valid: boolean;
  value: string | number;
  blur(): void;
  focus(): void;
  reset(): void;
  setDirty(dirty: boolean): void;
  innerRef: RefObject<HTMLDivElement>;
  inputRef: RefObject<HTMLInputElement | HTMLTextAreaElement>;
}

export declare interface TextFieldProps
  extends SpecialComponentPropsWithRef<any, TextFieldRef> {
  autoFocus?: boolean;
  disabled?: boolean;
  required?: boolean;
  type?: string;
  valid?: boolean;
  value?: string | number;
  onBlur?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onChange?(field: FieldContent<string | number>): void;
  onFocus?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onWheel?(event: WheelEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onValidate?(
    val: string | number,
    flags: { required: boolean; dirty: boolean }
  ): boolean;
}

export declare interface TextFieldState {
  value: string | number;
  valid: boolean;
  dirty: boolean;
  focused: boolean;
}

const TextField = ({
  ref,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val: string | number, opts?: { required: boolean }) =>
    !!val || !opts?.required,
  ...rest
}: TextFieldProps) => {
  const innerRef = useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLInputElement & HTMLTextAreaElement>(null);
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState<TextFieldState>, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = (value?: string) => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = (
    e: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = (dirty: boolean) => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? ((_: WheelEvent) => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
};

TextField.displayName = 'TextField';

export default TextField;
