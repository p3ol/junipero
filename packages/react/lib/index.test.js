import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
