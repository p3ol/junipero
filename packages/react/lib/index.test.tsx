import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
