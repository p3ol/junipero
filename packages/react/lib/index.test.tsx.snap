// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mockState: inside react > should correctly update the state and return the new state when using mockState with a callback > With updated content AND updated text field valud 1`] = `
<div>
  <div>
    <div
      class="junipero text-field dirty valid"
    >
      <input
        class="field"
        data-testid="Field"
        type="text"
        value="This is a test 2"
      />
    </div>
    <button>
      Add content
    </button>
    <div>
      Text: 
      This is a test 2
    </div>
    <div>
      Content: 
      foo, bar
    </div>
  </div>
</div>
`;

exports[`mockState: inside react > should correctly update the state and return the new state when using mockState with a callback > With updated text field value 1`] = `
<div>
  <div>
    <div
      class="junipero text-field dirty valid"
    >
      <input
        class="field"
        data-testid="Field"
        type="text"
        value="This is a test 2"
      />
    </div>
    <button>
      Add content
    </button>
    <div>
      Text: 
      This is a test 2
    </div>
    <div>
      Content: 
      foo
    </div>
  </div>
</div>
`;
